src/libNetwork/Whitelist.cpp

```diff
@@ -0,0 +1,116 @@
+/**
+* Copyright (c) 2018 Zilliqa
+* This source code is being disclosed to you solely for the purpose of your participation in
+* testing Zilliqa. You may view, compile and run the code for that purpose and pursuant to
+* the protocols and algorithms that are programmed into, and intended by, the code. You may
+* not do anything else with the code without express permission from Zilliqa Research Pte. Ltd.,
+* including modifying or publishing the code (or any part of it), and developing or forming
+* another public or private blockchain network. This source code is provided ‘as is’ and no
+* warranties are given as to title or non-infringement, merchantability or fitness for purpose
+* and, to the extent permitted by law, all liability for your use of the code is disclaimed.
+* Some programs in this code are governed by the GNU General Public License v3.0 (available at
+* https://www.gnu.org/licenses/gpl-3.0.en.html) (‘GPLv3’). The programs that are governed by
+* GPLv3.0 are those programs that are located in the folders src/depends and tests/depends
+* and which include a reference to GPLv3 in their program files.
+*
+* This should only be used in testnet release  only. This is to ensure the stability of testnet.
+* Mainnet will not require this function and nodes will be incentivise to perform the role
+* as member of DS committee.
+**/
+
+#include "Whitelist.h"
+
+#include <arpa/inet.h>
+#include <boost/foreach.hpp>
+#include <boost/property_tree/ptree.hpp>
+#include <boost/property_tree/xml_parser.hpp>
+#include <iostream>
+
+#include "Peer.h"
+#include "libUtils/DataConversion.h"
+#include "libUtils/Logger.h"
+
+using namespace std;
+using namespace boost::multiprecision;
+
+Whitelist::Whitelist() {}
+
+Whitelist::~Whitelist() {}
+
+Whitelist& Whitelist::GetInstance()
+{
+    static Whitelist whitelistInfo;
+    return whitelistInfo;
+}
+
+void Whitelist::AddToDSWhitelist()
+{
+    LOG_MARKER();
+
+    ifstream config("whitelist.xml");
+
+    if (!TEST_NET_MODE)
+    {
+        LOG_GENERAL(WARNING, "Not in testnet mode. Whitelisting not allowed");
+        return;
+    }
+
+    if (config.fail())
+    {
+        LOG_GENERAL(WARNING, "No whitelist xml present");
+        return;
```

Comment:
config.close() should be called here because ifstream constructor would open file.
http://www.cplusplus.com/reference/fstream/ifstream/ifstream/

Reply:
On a deeper look, close() is not necessary here as it failed to open file. Calling close() will result in exception here.

Reply:
ok

Reply:
It's not very necessary to call `close()` according to RAII.